lib/copy/copy.js

```diff
@@ -30,209 +26,188 @@ function copy (src, dest, opts, cb) {
   }
-  return setDestTimestampsAndMode(srcMode, src, dest, cb)
+  return setDestTimestampsAndMode(srcMode, src, dest)
 }
 
 function fileIsNotWritable (srcMode) {
   return (srcMode & 0o200) === 0
 }
 
-function makeFileWritable (dest, srcMode, cb) {
-  return setDestMode(dest, srcMode | 0o200, cb)
+function makeFileWritable (dest, srcMode) {
+  return setDestMode(dest, srcMode | 0o200)
 }
 
-function setDestTimestampsAndMode (srcMode, src, dest, cb) {
-  setDestTimestamps(src, dest, err => {
-    if (err) return cb(err)
-    return setDestMode(dest, srcMode, cb)
-  })
+async function setDestTimestampsAndMode (srcMode, src, dest) {
+  await setDestTimestamps(src, dest)
+  return setDestMode(dest, srcMode)
 }
 
-function setDestMode (dest, srcMode, cb) {
-  return fs.chmod(dest, srcMode, cb)
+function setDestMode (dest, srcMode) {
+  return fs.chmod(dest, srcMode)
 }
 
-function setDestTimestamps (src, dest, cb) {
+async function setDestTimestamps (src, dest) {
   // The initial srcStat.atime cannot be trusted
   // because it is modified by the read(2) system call
-  // (See https://nodejs.org/api/fs.html#fs_stat_time_values)
-  fs.stat(src, (err, updatedSrcStat) => {
-    if (err) return cb(err)
-    return utimesMillis(dest, updatedSrcStat.atime, updatedSrcStat.mtime, cb)
-  })
+  // (See https://nodejs.org/api/fs.html#fs$stat_time_values)
+  const updatedSrcStat = await fs.stat(src)
+
+  return utimesMillisAsync(dest, updatedSrcStat.atime, updatedSrcStat.mtime)
 }
 
-function onDir (srcStat, destStat, src, dest, opts, cb) {
-  if (!destStat) return mkDirAndCopy(srcStat.mode, src, dest, opts, cb)
-  return copyDir(src, dest, opts, cb)
+function onDir (srcStat, destStat, src, dest, opts) {
+  if (!destStat) return mkDirAndCopy(srcStat.mode, src, dest, opts)
+  return copyDir(src, dest, opts)
 }
 
-function mkDirAndCopy (srcMode, src, dest, opts, cb) {
-  fs.mkdir(dest, err => {
-    if (err) return cb(err)
-    copyDir(src, dest, opts, err => {
-      if (err) return cb(err)
-      return setDestMode(dest, srcMode, cb)
-    })
-  })
+async function mkDirAndCopy (srcMode, src, dest, opts) {
+  await fs.mkdir(dest)
+  await copyDir(src, dest, opts)
+
+  return setDestMode(dest, srcMode)
 }
 
-function copyDir (src, dest, opts, cb) {
-  fs.readdir(src, (err, items) => {
-    if (err) return cb(err)
-    return copyDirItems(items, src, dest, opts, cb)
-  })
+async function copyDir (src, dest, opts) {
+  const items = await fs.readdir(src)
+  return copyDirItems(items, src, dest, opts)
 }
 
-function copyDirItems (items, src, dest, opts, cb) {
+function copyDirItems (items, src, dest, opts) {
   const item = items.pop()
-  if (!item) return cb()
-  return copyDirItem(items, item, src, dest, opts, cb)
+  if (!item) return
+  return copyDirItem(items, item, src, dest, opts)
 }
 
-function copyDirItem (items, item, src, dest, opts, cb) {
+async function copyDirItem (items, item, src, dest, opts) {
```

Comment:
This logic needs to be rethought/reworked. We currently use a recursive `copyDirItems` call, which mutates `items`. This made sense to avoid the complexity of serially iterating a callback-based process, but is non-idiomatic in an async/await world. I'm thinking we use a `for` loop in `copyDir` to iterate over the items...I'll leave the exact details to you.

Reply:
Refactored this part in [`cf3aed7` (#1021)](https://github.com/jprichardson/node-fs-extra/pull/1021/commits/cf3aed7a1d413e329fa7edede323f2479f3901a0).

The `onDir` now uses a for loop to iterate every item inside the directory (from `await fs.readdir`), perform the pre-copy check (the `filter` option and the `stat.checkPathsAsync`), and then invoke the `getStatsAndPerformCopy` function (which would recursively call `onDir` again).